### ISSUE TITLE
Add an ability to override comparator function

### ISSUE THREAD

**Issue by vasilich6107** [AhmedLSayed9/dropdown_button2#296]
There are some cases when we have no ability to override the `==` operator for example in case when we get a list of objects from external source.
It would be nice to have an ability to override object comparator function which is used to compare objects

<img width="694" alt="image" src="https://github.com/AhmedLSayed9/dropdown_button2/assets/6721286/fb45b58c-6039-4823-a4af-1aa579ab1cfe">

**Comment by AhmedLSayed9** [comment]
You can override the `==` operator for the object itself.
or, use some unique value at the object like an `id` to be used as value.

**Comment by AhmedLSayed9** [comment]
@vasilich6107 
Can you check https://github.com/AhmedLSayed9/dropdown_button2/issues/127 and see if it's what you're looking for?

**Comment by AhmedLSayed9** [comment]
What value will this add over doing:
```dart
class _MyHomePageState extends State<MyHomePage> {
  final List<Item> items = const [
    Item('1', 'Item1'),
    Item('2', 'Item2'),
    Item('3', 'Item3'),
    Item('4', 'Item4'),
  ];
  final valueListenable = ValueNotifier<String?>(null);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: DropdownButtonHideUnderline(
          child: DropdownButton2<String>(
            isExpanded: true,
            hint: Text(
              'Select Item',
              style: TextStyle(
                fontSize: 14,
                color: Theme.of(context).hintColor,
              ),
            ),
            items: items
                .map((Item item) => DropdownItem<String>(
                      value: item.id,
                      height: 40,
                      child: Text(
                        item.title,
                        style: const TextStyle(
                          fontSize: 14,
                        ),
                      ),
                    ))
                .toList(),
            valueListenable: valueListenable,
            onChanged: (String? value) {
              valueListenable.value = value;
            },
            buttonStyleData: const ButtonStyleData(
              padding: EdgeInsets.symmetric(horizontal: 16),
              height: 40,
              width: 140,
            ),
          ),
        ),
      ),
    );
  }
}

class Item {
  const Item(this.id, this.title);

  final String id;
  final String title;
}
```

**Comment by Mir1001** [comment]
I am experiencing a similar issue. Specifically, I am retrieving a list of users from an API, which includes the currently logged-in user but with less detailed information (same ID, public data, etc.). When I try to set the selected item/object to the logged-in user, the object isn't found because the equality operator (==) is auto-generated by Freezed and I prefer not to modify it.

As a workaround, I could replace the logged-in user in the list if it exists or use just IDs as values, though this approach is less intuitive. What is best approach?

**Comment by AhmedLSayed9** [comment]
> I am experiencing a similar issue. Specifically, I am retrieving a list of users from an API, which includes the currently logged-in user but with less detailed information (same ID, public data, etc.). When I try to set the selected item/object to the logged-in user, the object isn't found because the equality operator (==) is auto-generated by Freezed and I prefer not to modify it.
> 
> As a workaround, I could replace the logged-in user in the list if it exists or use just IDs as values, though this approach is less intuitive. What is best approach?

What's wrong with using `id` as value here?

**Comment by Mir1001** [comment]
[AhmedLSayed9](https://github.com/AhmedLSayed9) Thanks. It is true. 

I was under the impression that the types for **items<Type1>** and **DropdownButton2<Type2>** needed to be the same.

```dart
...
        items: items //can be type of Item
                .map((Item item) => DropdownItem<String>( //type of key
                      value: item.id,
                      height: 40,
```

**Comment by AhmedLSayed9** [comment]
> In your example [#296 (comment)](https://github.com/AhmedLSayed9/dropdown_button2/issues/296#issuecomment-2178993745) `value` will be `id` so when I need to do something with the output from dropdown I need to search again through
> 
> ```
> final List<Item> items = const [
>     Item('1', 'Item1'),
>     Item('2', 'Item2'),
>     Item('3', 'Item3'),
>     Item('4', 'Item4'),
>   ];
> ```
> 
> to get the actual item being selected.
> 
> It would be easier to have comparison function.

Yes, you'll have to search by id to get the `Item` back but that's not a big deal even with huge lists.

Also, you can always override the `==` operator for the object so this is only needed when you're using an object from another api and you don't want to extend it.


> 
> Could you clarify how suggested approach works with sealed classes which should be compared by type.
> 
> What should I do if I have
> 
> ```
> sealed calss A {}
> 
> class B extends A {
> 
> }
> 
> class C extends A {
> 
> }
> ```
> 
> so I want my items to be
> 
> ```
> items: [C(), B(),]
> ```
> 
> The A and B classes does not have similar fields, they could be compared only by type.

I'm not sure if there's a real use-case like this but anyway, you can do:
```dart
sealed class A {
  const A();
}

class B extends A {
  const B();
}

class C extends A {
  const C();
}
```
```dart
items: const [C(), B(),]
```

**Comment by AhmedLSayed9** [comment]
> Could you be so kind and allow users to override comparison method?

The thing is this how the core [DropdownButton](https://api.flutter.dev/flutter/material/DropdownButton-class.html) works and it's more familiar to users.

Also, we don't want to add more apis and complicate things unless it's really needed.

I still can't find a strong reason for doing:
```dart
valueComparator: (value) => value.id,
```

rather than:
```dart
value: value.id,
```

The only benefit here is the user won't need to search back to get the actual item.

I'm thinking we might make it optional and keep `value` then user can only use one of `value` and `valueComperator` but this is repetitive for nothing, or, just update `value` to simplify things then user will have to do:
```
value: (value) => value,
```
which seems a bit weird.

That's why I'm a bit skeptical about this change.

**Comment by AhmedLSayed9** [comment]
Also, this will break the dropdown menu items consistency, as they use the same type specified for the DropdownButton.
```
/// The type `T` is the type of the value the entry represents. All the entries
/// in a given menu must represent values with consistent types.
```

This change will allow items to have different value types which is not allowed.

**Comment by AhmedLSayed9** [comment]
Furthermore, You'll have to set `valueListenable` that represents the value of the currently selected [DropdownItem] with the same value you set to the item. This means you'll still need to search back to get the actual item :)

**Comment by AhmedLSayed9** [comment]
> What about the example with sealed classes?

I've answered above.

**Comment by AhmedLSayed9** [comment]
@vasilich6107
Can you show an example of what you're trying to do? Maybe we're looking for a wrong solution after all (see https://xyproblem.info)
